pkg/asset/k8s.go

```diff
@@ -116,16 +118,38 @@ const validBootstrapTokenChars = "0123456789abcdefghijklmnopqrstuvwxyz"
 // newBootstrapToken constructs a bootstrap token in conformance with the following format:
 // https://kubernetes.io/docs/admin/bootstrap-tokens/#token-format
 func newBootstrapToken() (id string, secret string, err error) {
-	// Read 6 random bytes for the id and 16 random bytes for the token (see spec for details).
-	token := make([]byte, 6+16)
-	if _, err := rand.Read(token); err != nil {
+	idBytes := make([]byte, 6)
+	secretBytes := make([]byte, 16)
+	tokenLength := len(idBytes) + len(secretBytes)
+	charsetLength := len(validBootstrapTokenChars)
+	size := int(math.Ceil(float64(tokenLength) * math.Log2(float64(charsetLength)) / math.Log2(256)))
+
+	random := make([]byte, size)
+	if _, err := rand.Read(random); err != nil {
 		return "", "", err
 	}
 
-	for i, b := range token {
-		token[i] = validBootstrapTokenChars[int(b)%len(validBootstrapTokenChars)]
+	encodeBytes(validBootstrapTokenChars, random, idBytes, secretBytes)
+	return string(idBytes), string(secretBytes), nil
+}
+
+// encodeBytes encodes the input byte slice into encoded byte slices.
+// The input is treated as a big-endian unsigned integer, and the
+// encoded slices represent the input value in little-endian unsigned
```

Comment:
Confused why this is mentioning "little-endian"

Reply:
> Confused why this is mentioning "little-endian"

In the test, input bytes are big-endian, so `[255, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]` is 255 * 256^14 + 2 * 256^13 + ... + 12^256^3 +  13 * 256^2 + 14 * 256 + 15 = 1324076502674089023697701463353265679.

The output encoded value is little endian, so `n2r8j0ug4twvqmqt53hyff` is [23, 2, 27, 8, 19, 0, 30, 16, 4, 29, 32, 31, 26, 22, 26, 29, 5, 3, 17, 34, 15, 15] (decoding each character to decimal, e.g. `n` -> 23, `2` -> 2, ...) which is 23 + 2 * 36 + 27 * 36^2 + ... + 15 * 36^20 + 15 * 36^21 = 7431760576136592951846490044435983.  If you add on the next two unused values (which a 24-character token would have consumed), you get 7431760576136592951846490044435983 + 4 * 36^22 + 2 * 36^23 = 1324076502674089023697701463353265679, the input number.

So the input is being interpreted as a big-endian unsigned integer, and the encoded output is little-endian base `len(charset)`.  If we had big-endian output, it would have been `ffyh35tqmqvwt4gu0j8r2n` instead of `n2r8j0ug4twvqmqt53hyff`.